loop invariant code motion

o- opened the pull request.

o- committed:
loop invariant code motion

charig commented on rir/src/compiler/opt/hoist_instruction.cpp:
```diff
@@ -0,0 +1,131 @@
+
+                // !trg->isBranch, is a cheap heuristic to avoid moving
+                // something from the loop body to the loop header, which is
+                // usually not great.
```
Why not great to move to the loop header? Where does the loop concept actually appear in all these code? Why we assume a loop header is branch node?

charig commented on rir/src/compiler/opt/hoist_instruction.cpp:
```diff
@@ -0,0 +1,131 @@
+                // Try to find a hoisting candidate that is dominated by all
+                // arguments to i
+                // cppcheck-suppress knownConditionTrueFalse
+                if (!trg)
```
target?

charig commented on rir/src/compiler/opt/pass_definitions.h:
```diff
@@ -133,6 +133,8 @@ class PASS(TypeInference);
 
 class PASS(TypeSpeculation);
 
+class PASS(HoistInstruction);
```
Would you mind adding a comment specifying exactly what does this pass do?

o- replied:
to have it become outdated like the others :)

charig commented on rir/src/utils/configurations.cpp:
```diff
@@ -51,6 +51,7 @@ void Configurations::defaultOptimizations() {
         optimizations.push_back(new pir::Constantfold());
         optimizations.push_back(new pir::Cleanup());
         optimizations.push_back(new pir::DelayInstr());
+        optimizations.push_back(new pir::HoistInstruction());
```
Maybe it is time to change the name of DelayInstr and also maybe be more specific with HoistInstruction? For instance, `DelayLookup`, `HoistEffectless`?

o- replied:
it's not HoistEffectless. yes I agree that we should review the passes, also check if they actually work, remove stupid ones, cleanup naming and so on. but not really related to this pr

charig commented on rir/src/compiler/opt/hoist_instruction.cpp:
```diff
@@ -0,0 +1,131 @@
+
+                std::function<bool(BB*)> doesNotReorderEffects = [&](BB* x) {
+                    // empty or loop
+                    if (!x || x == trg)
```
I do not get it. How do you distinguish the first iteration from a loop iteration? How are you sure that there is no another loop that does not start at trg?

o- committed:
more reasonable heuristic when to hoist

check if we duplicate computation by hoisting

o- commented:
@charig I changed some stuff, addressed your comments and added some more documentation. feel free to request more changes still. it's not a great optimization. it does not work for instructions without inputs (for example it does not work for LdVar). but it does the trick for hoisting forces out of loops in many occasions...

charig commented on rir/src/compiler/opt/hoist_instruction.cpp:
```diff
@@ -0,0 +1,187 @@
+                        // waste computation
+                        if (x->last()->branches()) {
+                            if (!dom.dominates(x->trueBranch(), bb) ||
+                                !dom.dominates(x->falseBranch(), bb)) {
```
Why there is one exception?

o- replied:
supposedly a loop is something like:

```
a:
  branch b c
b:
    ....
    goto a
c:
```

so to hoist an instruction from `...` to `a:` we need to cross one branch instructions where not both branch targets dominate the loop body.

this is basically a very localized formulation of, only hoist out of one loop.

o- replied:
as I said in the comment, the better solution would be, to compute loop headers. then explicitly hoist instructions there. this would also allow us to hoist instructions without arguments. here I am just using the positions of the arguments of an instruction as a cheap heuristic of where it could be profitable to place an instruction.

maybe this could be achieved by combining the changes here with #362

charig commented on rir/src/compiler/opt/hoist_instruction.cpp:
```diff
@@ -0,0 +1,187 @@
+            bool success = true;
+            if (i->hasObservableEffects())
+                success = doesNotReorderEffects(target);
+            else if (i->cost() > 0)
```
Why we have a blacklist but then a cost for ruling out `castType`?

o- replied:
it's the other way round. castType is free (since it's not an actual instruction) so we can always hoist it (success is always true)

charig commented on rir/src/compiler/pir/instruction.h:
```diff
@@ -937,6 +938,7 @@ class FLIE(Force, 2, Effects::Any()) {
 
 class FLI(CastType, 1, Effects::None()) {
   public:
+    unsigned cost() const override final { return 0; }
```
cost of what?

o- replied:
idk :) how expensive to execute, something... currently my idea was just to distinguish between instructions which will be scheduled in rir, and "virtual" ones.

but the idea was that maybe we can extend on it and use it for other optimizations. e.g. an add with environment could be more expensive than plain add. but I am not really sure. I would grow this as we need it...
